cmd/agent/main.go

```diff
@@ -1,3 +1,224 @@
 package main
 
-func main() {}
+import (
+	"flag"
+	"fmt"
+	"log"
+	"math/rand"
+	"net/http"
+	"os"
+	"os/signal"
+	"runtime"
+	"strconv"
+	"strings"
+	"time"
+)
+
+const (
+	defaultServerAddress  = "http://localhost:8080"
+	defaultPollInterval   = 2
+	defaultReportInterval = 10
+)
+
+var (
+	serverAddress  string
+	pollInterval   time.Duration
+	reportInterval time.Duration
+	pollCount      int64
+)
+
+// List of metrics to collect from runtime.MemStats
+var gaugeMetrics = []string{
+	"Alloc", "BuckHashSys", "Frees", "GCCPUFraction", "GCSys", "HeapAlloc",
+	"HeapIdle", "HeapInuse", "HeapObjects", "HeapReleased", "HeapSys",
+	"LastGC", "Lookups", "MCacheInuse", "MCacheSys", "MSpanInuse", "MSpanSys",
+	"Mallocs", "NextGC", "NumForcedGC", "NumGC", "OtherSys", "PauseTotalNs",
+	"StackInuse", "StackSys", "Sys", "TotalAlloc",
+}
+
+func main() {
+	// Read flags
+	flagAddress := flag.String("a", "", "HTTP server address (default: http://localhost:8080)")
+	flagReport := flag.Int("r", 0, "Report interval in seconds (default: 10)")
+	flagPoll := flag.Int("p", 0, "Poll interval in seconds (default: 2)")
+	flag.Parse()
+
+	if len(flag.Args()) > 0 {
+		log.Fatalf("Unknown flags: %v", flag.Args())
+	}
+
+	// --- Address
+	address := os.Getenv("ADDRESS")
+	if address == "" {
+		if *flagAddress != "" {
+			address = *flagAddress
+		} else {
+			address = defaultServerAddress
+		}
+	}
+	serverAddress = address
+
+	if !strings.HasPrefix(serverAddress, "http://") && !strings.HasPrefix(serverAddress, "https://") {
+		serverAddress = "http://" + serverAddress
+	}
+	// --- Report Interval
+	reportEnv := os.Getenv("REPORT_INTERVAL")
+	var reportSeconds int
+	if reportEnv != "" {
+		val, err := strconv.Atoi(reportEnv)
+		if err != nil {
+			log.Fatalf("Invalid REPORT_INTERVAL: %v", err)
+		}
+		reportSeconds = val
+	} else if *flagReport != 0 {
+		reportSeconds = *flagReport
+	} else {
+		reportSeconds = defaultReportInterval
+	}
+	reportInterval = time.Duration(reportSeconds) * time.Second
+
+	// --- Poll Interval
+	pollEnv := os.Getenv("POLL_INTERVAL")
+	var pollSeconds int
+	if pollEnv != "" {
+		val, err := strconv.Atoi(pollEnv)
+		if err != nil {
+			log.Fatalf("Invalid POLL_INTERVAL: %v", err)
+		}
+		pollSeconds = val
+	} else if *flagPoll != 0 {
+		pollSeconds = *flagPoll
+	} else {
+		pollSeconds = defaultPollInterval
+	}
+	pollInterval = time.Duration(pollSeconds) * time.Second
+
+	// --- Main program starts
+	gauges := make(map[string]float64)
+
+	tickerPoll := time.NewTicker(pollInterval)
+	tickerReport := time.NewTicker(reportInterval)
+	defer tickerPoll.Stop()
+	defer tickerReport.Stop()
```

Comment:
Стоит вызывать `defer` непосредственно после инициализации структуры, которую будет очищать.
```suggestion
	tickerPoll := time.NewTicker(pollInterval)
	defer tickerPoll.Stop()
	tickerReport := time.NewTicker(reportInterval)
	defer tickerReport.Stop()
```